Source/JavaScriptCore/dfg/DFGConstantFoldingPhase.cpp

```diff
@@ -1094,6 +1094,28 @@ class ConstantFoldingPhase : public Phase {
 
                 break;
             }
+
+            case HasStructureWithFlags: {
+                const AbstractValue& abstractValue = m_state.forNode(node->child1());
+                unsigned flags = node->structureFlags();
+                ASSERT(flags);
+
+                if (JSValue value = abstractValue.value()) {
+                    m_graph.convertToConstant(node, jsBoolean(asObject(value)->structure()->bitField() & flags));
+                    break;
+                }
+
+                if (abstractValue.m_structure.isFinite()) {
```

Comment:
Let's first check that `abstractValue.m_type && !(abstractValue.m_type & ~SpecCell)`